### PR TITLE
BAR Hotkeys: Use txt format and keyload/keyreload

### PR DESCRIPTION
**Pull request by badosu**
With the addition of keyload/keyreload on uikeys with arguments and
keysave with argument now we can safely switch back to .txt and make bar
hotkeys finally compatible with uikeys.txt

Users that previously had to disable bar hotkeys now can simply use
their uikeys with the preset custom enabled.

To load engine defaults simply put add `keydefaults` to the top of the file.

Users of custom preset on older format seamlessly migrate to new format, being written to `uikeys_bar.txt`.

This simplifies the format.

### CODE REVIEW

**Comment by badosu** [comment]
This PR is ready for review.

**Comment by badosu** [comment]
> Can you describe what needs to be tested here? I'm probably not qualified to review the code but I can help verify that everything is working

- Switching presets should work as before
- Migrating from previous version to this one retains previous preset config and loads same keybindings as before
- If user had previously used custom preset, it should have migrated from `bar_hotkeys_custom.lua` to an equivalent `uikeys_bar.txt` automatically on the same folder.

To test keybindings are equivalent one can `/keysave kb1.txt` on before version, then load updated version and `/keysave kb2.txt`, both kb1.txt and kb2.txt files should be equivalent except for order

**Comment by Beherith** [comment]
@thehobojoe , have you perchance had an opportunity to look at this one?

**Comment by thehobojoe** [comment]
> @thehobojoe , have you perchance had an opportunity to look at this one?

No I got delayed, I am starting to test it now. There's a lot of ground to cover so it may take a little while

**Comment by thehobojoe** [comment]
Ok I have tested a fair bit so far, nothing is broken but there is some weirdness.

Conversion from custom lua file to custom uikeys file worked great. However the generated file is called `uikeys_bar_user.txt` which is a mouthful and also disagrees with the echos that tell you to do `/keyreload uikeys_bar.txt`. After changing presets a few times I also got a `uikeys_bar.txt` file autogenerated, so now I have both in the same directory and I'm unsure why. Is there a reason the generated file doesn't always get called `uikeys.txt`? It seems like it would be more straightfoward, and it also means the user doesn't have to type `/keyreload filename` every time, since `uikeys.txt` is automatically checked for that command. I'm also unsure what happens when you have multiple uikeys files. Which one takes priority?

**Comment by badosu** [comment]
`uikeys_bar_user.txt` is just a backup we save on every game load, basically a snapshot of what the user had before last game, in case they need to revert it or not. It does not need to exist and perhaps is not aptly named.

`uikeys_bar.txt` is what we use for custom keys, only written if you actually have set custom preset.

`uikeys.txt` we allow it to exist independently of what we do on bar. Mostly for compatibility with people playing games other than bar with the same engine, but in theory we could use `uikeys.txt` and not care about this.

**Comment by badosu** [comment]
`/keyreload should work automatically for uikeys_bar.txt`

That is only possible if we intercept the `keyreload` action so it actually does `keyreload uikeys_bar.txt`, that's fine by me

File: LICENSE.md [hunk 1]
```diff
@@ -1,5 +1,5 @@
 The Code:
-The Spring Engine (which this game runs on) requires that all the code the games use are under the GPL licence. (https://springrts.com)
+The Spring Engine (which this game runs on) requires that all the code the games use are under the GPL v2 licence. (https://springrts.com)
```
Comment (sprunk) [inline LICENSE.md]: It only requires the code to be compatible with v2 or later. v3 is fine, MIT is fine, etc

**Comment by thehobojoe** [comment]
> That is only possible if we intercept the `keyreload` action so it actually does `keyreload uikeys_bar.txt`, that's fine by me.

This sounds good, as long as it continues to work with `uikeys.txt` as well (a lot of players use that file atm, including myself). Hopefully that is not too difficult to do.

Everything else seems good, my only requests before merge are they /keyreload thing and changing the name of `uikeys_bar_user.txt`. I'll do some more testing to verify behavior as well.

**Comment by thehobojoe** [comment]
@badosu any update on this? I've spent a decent amount of time in lua now so if you want to hand this off I think I can make any further changes needed.

**Comment by badosu** [comment]
I can't give any estimate right now, I trust you have the ability to adapt it to best suit the needs of the project.

Basically this is a critical week for me, and next one I'll have to catch up with work as well. So if you want to put your hands into this feel free, you can fork from it and send a new PR or just base your work on this one.

The most important part is just to remember switching to each preset on current keybindings (they have changed since then, maybe just grid and grid 60) and do `/keysave grid.txt` to use as the base for the new preset bindings file. Then for grid 60 you'd switch to it with the most up-to-date bindings and do `/keysave grid_60.txt` for example and so on.

**Pull request merged**